akka-stream/src/main/scala/akka/stream/io/Framing.scala

```diff
@@ -81,59 +81,42 @@ object Framing {
   def simpleFramingProtocol(maximumMessageLength: Int): BidiFlow[ByteString, ByteString, ByteString, ByteString, NotUsed] = {
     val decoder = lengthField(4, 0, maximumMessageLength + 4, ByteOrder.BIG_ENDIAN).map(_.drop(4))
     val encoder = Flow[ByteString].transform(() ⇒ new PushStage[ByteString, ByteString] {
-
-      override def onPush(message: ByteString, ctx: Context[ByteString]): SyncDirective = {
-        if (message.size > maximumMessageLength)
-          ctx.fail(new FramingException(s"Maximum allowed message size is $maximumMessageLength " +
-            s"but tried to send ${message.size} bytes"))
-        else {
-          val header = ByteString(
-            (message.size >> 24) & 0xFF,
-            (message.size >> 16) & 0xFF,
-            (message.size >> 8) & 0xFF,
-            message.size & 0xFF)
+      override def onPush(message: ByteString, ctx: Context[ByteString]): SyncDirective = message.size match {
+        case sz if sz > maximumMessageLength ⇒
+          ctx.fail(new FramingException(s"Maximum allowed message size is $maximumMessageLength but tried to send $sz bytes"))
+        case sz ⇒
+          val header = ByteString((sz >> 24) & 0xFF, (sz >> 16) & 0xFF, (sz >> 8) & 0xFF, sz & 0xFF)
           ctx.push(header ++ message)
-        }
       }
-
     })
 
     BidiFlow.fromFlowsMat(encoder, decoder)(Keep.left)
   }
 
-  private trait IntDecoder {
-    def decode(bs: ByteIterator): Int
-  }
-
   class FramingException(msg: String) extends RuntimeException(msg)
 
-  private class BigEndianCodec(val length: Int) extends IntDecoder {
-    override def decode(bs: ByteIterator): Int = {
-      var count = length
-      var decoded = 0
-      while (count > 0) {
-        decoded <<= 8
-        decoded |= bs.next().toInt & 0xFF
-        count -= 1
-      }
-      decoded
+  private final val bigEndianDecoder: (ByteIterator, Int) ⇒ Int = (bs, length) ⇒ {
+    var count = length
+    var decoded = 0
+    while (count > 0) {
+      decoded <<= 8
+      decoded |= bs.next().toInt & 0xFF
+      count -= 1
     }
+    decoded
   }
 
-  private class LittleEndianCodec(val length: Int) extends IntDecoder {
-    private val highestOctet = (length - 1) * 8
-    private val Mask = ((1L << (length * 8)) - 1).toInt
-
-    override def decode(bs: ByteIterator): Int = {
-      var count = length
-      var decoded = 0
-      while (count > 0) {
-        decoded >>>= 8
-        decoded += (bs.next().toInt & 0xFF) << highestOctet
-        count -= 1
-      }
-      decoded & Mask
+  private final val littleEndianDecoder: (ByteIterator, Int) ⇒ Int = (bs, length) ⇒ {
+    val highestOctet = (length - 1) << 3
+    val Mask = ((1L << (length << 3)) - 1).toInt
```

Comment:
Given the prevalence of bitshifts in the existing code, switching to explicit shifting for * 8 makes sense to me :)